ballerina/validator_visitor.bal

```diff
@@ -367,25 +350,62 @@ class ValidatorVisitor {
+                string message = string `${expectedTypeName} cannot represent non ${expectedTypeName} value: []`;
                 ErrorDetail errorDetail = getErrorDetailRecord(message, location);
                 self.errors.push(errorDetail);
             }
         } else {
             string expectedTypeName = getOfType(inputValue.'type).name.toString();
             string value = variableValues.toString();
-            string message = string`${expectedTypeName} cannot represent non ${expectedTypeName} value: ${value}`;
+            string message = string `${expectedTypeName} cannot represent non ${expectedTypeName} value: ${value}`;
             ErrorDetail errorDetail = getErrorDetailRecord(message, location);
             self.errors.push(errorDetail);
         }
     }
 
-    isolated function coerceArgumentIntValueToFloat(parser:ArgumentNode argNode, __InputValue schemaArg) {
+    isolated function coerceArgumentNodeValue(parser:ArgumentNode argNode, __InputValue schemaArg) {
         string expectedTypeName = getOfType(schemaArg.'type).name.toString();
-        if expectedTypeName == FLOAT && getTypeName(argNode) == INT {
-            argNode.setValue(<float> argNode.getValue());
-            argNode.setKind(parser:T_FLOAT);
+        if argNode.isVariableDefinition() && argNode.getVariableValue() is Scalar {
+            Scalar value = <Scalar>argNode.getVariableValue();
+            value = self.coerceValue(value, expectedTypeName, getTypeNameFromValue(value), argNode.getValueLocation());
+            argNode.setVariableValue(value);
+            if value is decimal|float {
+                argNode.setKind(parser:T_FLOAT);
+            }
+        } else if argNode.getValue() is Scalar {
+            Scalar value = <Scalar>argNode.getValue();
+            value = self.coerceValue(value, expectedTypeName, getTypeNameFromValue(value), argNode.getValueLocation());
+            argNode.setValue(value);
+            if value is decimal|float {
+                argNode.setKind(parser:T_FLOAT);
+            }
+        }
+    }
+
+    isolated function coerceValue(Scalar value, string expectedTypeName, string actualTypeName, Location location)
+        returns Scalar {
+        if expectedTypeName == FLOAT && (actualTypeName == INT || actualTypeName == DECIMAL) {
+            float|error coerceValue = float:fromString(value.toString());
+            if coerceValue is float {
+                return coerceValue;
+            } else {
+                string message = string `${expectedTypeName} cannot represent non ${expectedTypeName} value: ` +
+                string `${value.toString()}`;
+                ErrorDetail errorDetail = getErrorDetailRecord(message, location);
+                self.errors.push(errorDetail);
+            }
+        } else if expectedTypeName == DECIMAL && (actualTypeName == INT || actualTypeName == FLOAT) {
```

Comment:
From both `if` and `else` blocks, `actualTypeName == INT` in common. Is there a way to take it out? That will reduce the complexity and increase the readability of the code.

Or maybe we can simplify the logic using boolean transformations.

Reply:
Taking `actualTypeName == INT` out  may increase the size of the function. Any suggestion for a simplified logic? or we can add an inner `if` block like 
`else if expectedTypeName == DECIMAL {
            if actualTypeName == INT || actualTypeName == FLOAT {`

Reply:
This looks good.